### PR TITLE
feat(transport): Add server graceful shutdown

### PR DESCRIPTION
**Pull request by xd009642**
Some error messages, one about hyper::server::Builder not existing and another about lack of clone on futures. Just putting this up as an initial WIP to make sure I'm on the right track :)

## Motivation

Recommended good practice for gRPC so should be exposed

## Solution
Expose it via builder pattern

### CODE REVIEW

File: tonic/src/transport/server.rs [hunk 1]
```diff
@@ -110,6 +111,14 @@ impl Server {
         }
     }
 
+    /// Set the graceful shutdown thing
+    pub fn graceful_shutdown(self, shutdown: Box<dyn Future<Output=()>>) -> Self {
```
Comment (LucioFranco) [inline tonic/src/transport/server.rs]: I'd probably make this `impl Future<Output = ()>` and then box it internally.

File: tonic/src/transport/server.rs [hunk 2]
```diff
@@ -268,15 +289,19 @@ impl Server {
             concurrency_limit,
             // timeout,
         };
```
Comment (LucioFranco) [inline tonic/src/transport/server.rs]:
```suggestion
        };

```

**Comment by LucioFranco** [comment]
hmm looks like there are some compile errors?

**Comment by xd009642** [comment]
Yeah `hyper::server::builder` isn't found in `hyper::server` but that's not something I changed...

**Comment by xd009642** [comment]
Oh wait I did change something! Sorting it now

**Comment by xd009642** [comment]
@LucioFranco did you have a chance to look at my type based woes?

**Comment by LucioFranco** [comment]
Looks like there are a couple of CI failures, mind fixing them, then we can merge this.

**Comment by xd009642** [comment]
Yeah but I'm not sure quite how, my attempt to make serve_common to avoid repeating code leads to ```type inside `async` object must be known in this context``` because I think the acceptor type isn't known from the calling context. I could just duplicate the code across the two functions but I was hoping to do something cleaner than that

**Pull request merged**